draft-ietf-mls-protocol.md

```diff
@@ -505,47 +594,45 @@ A              B              C          Directory            Channel
 |  Welcome(B)  |              |              |                   |
-|------------->|state.join()  |              |                   |
+|------------->|              |              |                   |
 |              |              |              |                   |
 |              |              |              | Add(A->AB)        |
 |              |              |              | Commit(Add)       |
 |<---------------------------------------------------------------|
-|state.add(B)  |              |              |                   |
 |              |              |              |                   |
 |              |              |              |                   |
 |              |              |              | Add(AB->ABC)      |
 |              |              |              | Commit(Add)       |
 |--------------------------------------------------------------->|
 |              |              |              |                   |
 |              |  Welcome(C)  |              |                   |
-|---------------------------->|state.join()  |                   |
+|---------------------------->|              |                   |
 |              |              |              |                   |
 |              |              |              | Add(AB->ABC)      |
 |              |              |              | Commit(Add)       |
 |<---------------------------------------------------------------|
-|state.add(C)  |<------------------------------------------------|
-|              |state.add(C)  |              |                   |
+|              |<------------------------------------------------|
 |              |              |              |                   |
 ~~~~~
+{: title="Client A creates a group with clients B and C"}
 
 Subsequent additions of group members proceed in the same way.  Any
 member of the group can download a KeyPackage for a new client
-and broadcast an Add message that the current group can use to update
+and broadcast Add and Commit message that the current group can use to update
 their state, and a Welcome message that the new client can use to
 initialize its state and join the group.
 
-To enforce the forward secrecy and post-compromise security of messages,
-each member periodically updates their leaf secret.
-Any member can update this information at any time by generating a fresh
-KeyPackage and sending an Update message followed by a Commit message.
-Once all members have processed both, the group's secrets will be unknown to an
+To enforce the forward secrecy and post-compromise security of messages, each
+member periodically updates their leaf secret.  A member can update this message
+by sending a Commit (possibly with no proposals), or by sending an Update
+message that is committed by another member.  Once the other members of the
+group have processed these messages, the group's secrets will be unknown to an
 attacker that had compromised the sender's prior leaf secret.
```

Comment:
I think the core idea of *ratchet tree* is the same as the *sender key*. i.e., when committing a new group change, the committer will re-sample a new group secret $gs$ for the group. But the difficult part is how do the committer efficiently distributes $gs$ to other group members. In *sender key*, the committer simply encrypts $gs$ $N$ times (where N is the number of group members in this group) using each member's public key and delivers each $gs$'s ciphertext to its corresponding group member. In *ratchet tree*, group members are organized into a tree and each member is associated with a leaf node in the tree (i.e., there are $N$ leaf nodes). Furthermore, *ratchet tree* ensures that the secret key of a non-leaf node is known to the members whose corresponding leaf nodes are in the subtree rooted at that non-leaf node. In particular, the secret key of the root node is known by all group members. When the committer needs to distribute the newly sampled group secret $gs$ to the leaf nodes under a non-leaf node $k$, the committer can use the public key of $k$ to encrypt $gs$ instead of individually encrypting $gs$ using each leaf node's public keys. In this way, the ratchet tree achieves a more efficient group key distribution.

Reply:
In ratchet tree, the the group secret $gs$ can be viewed as the `secret` of the root node. To sample a new group secret (i.e., the root node secret $rs$) for the group, the committer first updates the ratchet tree by resampling a new leaf secret for its corresponding leaf node. With the updated ratchet tree, the root node secret $rs$ is updated. When distributing this updated ratchet tree to other group members, the committer needs to keep the `tree invariant`, that is, `the secret of a non-leaf node $k$ should be known by all leaf nodes under $k$`.

Reply:
To remove a member $m$, the committer simply does not send any information of the updated ratchet tree to $m$, so that $m$ won't know the updated group secret $gs$. 
For post-compromise security, that is, for a client $s$ to quickly recover from a compromised state (e.g., its corresponding leaf secret $ls$ is leaked to an attacker), the client $c$ can send a Commit message asking for resampling its leaf secret as $ls'$. As long as the updated leaf secret $ls'$ is unknown to the attacker, the attacker cannot learn any information regarding the newly updated group secret $gs'$.

Reply:
The differences to sender keys are larger than you say.  With sender keys, each sender has an independent key `k_i`; there is no shared group state.  To remove someone (including PCS updates), each sender needs to distribute a new key `k'_i` to the remaining `N-1` participants.  This results in `(N-1)^2` messages of size `O(1)`.

In MLS, because there is shared group state, you only need some entropy that is not sent to the evicted member, as you say.  So you only need a single message of size `O(N-1)` in the worst case and `O(log(N-1))` with a populated tree.